File: androidApp/src/main/java/com/mamba/picme/domain/organize/ConfidenceGrader.kt
package com.mamba.picme.domain.organize

/**
 * 置信度分级（纯函数）：决定详情页三段分组与默认勾选口径。
 * HIGH = 强信号命中（默认勾选）；MEDIUM = 单信号接近阈值（仅列出）；
 * LOW = 弱信号（美学分等辅助信号单独命中）。
 * 前置：item 已经 CategoryArbiter 裁定入 [category]，本层只评估信号强度。
 */
object ConfidenceGrader {

    fun grade(item: OrganizeItem, category: OrganizeCategory): OrganizeConfidence =
        when (category) {
            OrganizeCategory.DUPLICATES ->
                if (item.exactDupGroupSize >= 2) OrganizeConfidence.HIGH else OrganizeConfidence.MEDIUM

            // 路径判定为确定性强信号
            OrganizeCategory.SCREEN_CONTENT -> OrganizeConfidence.HIGH

            OrganizeCategory.DOCUMENTS -> gradeDocument(item)

            OrganizeCategory.LOW_QUALITY_PORTRAITS -> {
                val faceQuality = item.faceQualityScore ?: 1.0f
                if (faceQuality < OrganizeThresholds.FACE_QUALITY_LOW * OrganizeThresholds.STRONG_SIGNAL_FACTOR) {
                    OrganizeConfidence.HIGH
                } else {
                    OrganizeConfidence.MEDIUM
                }
            }

            OrganizeCategory.LOW_QUALITY_PHOTOS -> gradeLowQualityPhoto(item)

            OrganizeCategory.LARGE_FILES ->
                if (item.sizeBytes >= OrganizeThresholds.LARGE_FILE_STRONG_FACTOR * largeFileThreshold(item)) {
                    OrganizeConfidence.HIGH
                } else {
                    OrganizeConfidence.MEDIUM
                }
        }

    /** OCR 强命中（≥2×密度阈值）HIGH；OCR 命中或仅标签命中 MEDIUM。 */
    private fun gradeDocument(item: OrganizeItem): OrganizeConfidence {
        val chars = item.ocrText?.length ?: 0
        if (chars == 0) return OrganizeConfidence.MEDIUM // labels 关键词命中（无 OCR 佐证）
        val area = item.pixelArea
        val strongThreshold = if (area != null && area > 0) {
            // 与 DedupContentTypeDetector 同一面积归一口径的 2×
            area / 1_000_000L * OrganizeThresholds.OCR_DENSITY_PER_MEGAPIXEL * OrganizeThresholds.OCR_STRONG_FACTOR
        } else {
            OrganizeThresholds.OCR_DENSITY_FALLBACK_CHARS.toLong() * OrganizeThresholds.OCR_STRONG_FACTOR
        }
        return if (chars.toLong() >= strongThreshold) {
            OrganizeConfidence.HIGH
        } else {
            OrganizeConfidence.MEDIUM
        }
    }

    private fun gradeLowQualityPhoto(item: OrganizeItem): OrganizeConfidence {
        val blur = item.blurScore
        if (blur != null && blur < OrganizeThresholds.BLUR_VARIANCE_LOW * OrganizeThresholds.STRONG_SIGNAL_FACTOR) {
            return OrganizeConfidence.HIGH
        }
        val blurred = blur != null && blur < OrganizeThresholds.BLUR_VARIANCE_LOW
        val exposure = item.exposureScore
        val badExposure = exposure != null &&
            (exposure < OrganizeThresholds.EXPOSURE_UNDER || exposure > OrganizeThresholds.EXPOSURE_OVER)
        if (blurred || badExposure) return OrganizeConfidence.MEDIUM
        // 模糊/曝光正常或未计算，仅 NIMA 低分 → 弱信号
        return OrganizeConfidence.LOW
    }

    private fun largeFileThreshold(item: OrganizeItem): Long =
        if (item.isVideo) {
            OrganizeThresholds.LARGE_VIDEO_BYTES
        } else {
            OrganizeThresholds.LARGE_PHOTO_BYTES
        }
}

File: iosApp/PoLang/Features/Organize/Domain/ConfidenceGrader.swift
import Foundation

/// 置信度分级（纯函数）：决定详情页三段分组与默认勾选口径。
/// HIGH = 强信号命中（默认勾选）；MEDIUM = 单信号接近阈值（仅列出）；
/// LOW = 弱信号（美学分等辅助信号单独命中）。
/// 直译 Android domain/organize/ConfidenceGrader.kt。
/// 前置：item 已经 CategoryArbiter 裁定入 category，本层只评估信号强度。
enum ConfidenceGrader {

    static func grade(_ item: OrganizeItem, category: OrganizeCategory) -> OrganizeConfidence {
        switch category {
        case .duplicates:
            return item.exactDupGroupSize >= 2 ? .high : .medium

        // 路径/子类型判定为确定性强信号
        case .screenContent:
            return .high

        case .documents:
            return gradeDocument(item)

        case .lowQualityPortraits:
            let faceQuality = item.faceQualityScore ?? 1.0
            return faceQuality < OrganizeThresholds.faceQualityLow * OrganizeThresholds.strongSignalFactor
                ? .high : .medium

        case .lowQualityPhotos:
            return gradeLowQualityPhoto(item)

        case .largeFiles:
            return item.sizeBytes >= OrganizeThresholds.largeFileStrongFactor * largeFileThreshold(item)
                ? .high : .medium
        }
    }

    /// OCR 强命中（≥2×密度阈值）HIGH；OCR 命中或仅标签命中 MEDIUM。
    private static func gradeDocument(_ item: OrganizeItem) -> OrganizeConfidence {
        let chars = Int64(item.ocrText?.count ?? 0)
        if chars == 0 { return .medium } // labels 关键词命中（无 OCR 佐证）
        if let area = item.pixelArea, area > 0 {
            // 与 Android 同一面积归一口径的 2×（先乘后除，避免 sub-MP 整数截断塌缩为 0）
            return chars * 1_000_000 >= area * OrganizeThresholds.ocrDensityPerMegapixel * OrganizeThresholds.ocrStrongFactor
                ? .high : .medium
        }
        return chars >= OrganizeThresholds.ocrDensityFallbackChars * OrganizeThresholds.ocrStrongFactor
            ? .high : .medium
    }

    /// 模糊强命中 HIGH；模糊/曝光任一边界命中 MEDIUM；其余（含 nil 信号）LOW。
    private static func gradeLowQualityPhoto(_ item: OrganizeItem) -> OrganizeConfidence {
        if let blur = item.blurScore,
           blur < OrganizeThresholds.blurVarianceLow * OrganizeThresholds.strongSignalFactor {
            return .high
        }
        let blurred = (item.blurScore ?? .infinity) < OrganizeThresholds.blurVarianceLow
        let badExposure: Bool
        if let exposure = item.exposureScore {
            badExposure = exposure < OrganizeThresholds.exposureUnder || exposure > OrganizeThresholds.exposureOver
        } else {
            badExposure = false
        }
        if blurred || badExposure { return .medium }
        // 防御性兜底：arbiter 不定类 NIMA-only 项，正常管线不可达；nil 信号落最低档方向安全
        return .low
    }

    /// 大文件主阈值：视频按字节，照片按字节（像素面积门槛已在 CategoryArbiter 裁定）。
    private static func largeFileThreshold(_ item: OrganizeItem) -> Int64 {
        item.isVideo ? OrganizeThresholds.largeVideoBytes : OrganizeThresholds.largePhotoBytes
    }
}
